### DISCUSSION TITLE
Make the YOLOv8 LibTorch CPP Inference give filtered detections realtime

### DISCUSSION THREAD

**Discussion by shikhindahikar** [ultralytics/ultralytics#21073] [Q&A] [answered]
Hi, I am using [this example](https://github.com/ultralytics/ultralytics/tree/v8.3.114/examples/YOLOv8-LibTorch-CPP-Inference) given in the repo of ultralytics. I have modified it to ingest video stream and perform inference on the frames one by one, the issue is that in the loop where I am doing inference on the frame, I am clearly getting a huge slowdown because of NMS.

I always receive the total of 33600 raw detections only from the inference(which can give me good 50-300 FPS without NMS) 

`torch::Tensor output = yolo_model.forward(inputs).toTensor();   // Inference`

So I can either do a [CUDA based NMS](https://github.com/pytorch/vision/blob/main/torchvision/csrc/ops/cuda/nms_kernel.cu) which gives me filtered detections without overlapping bounding boxes(1-3 FPS) or if I just do a simpler filter using tensor functions like following
```
// filtering all the detections with less than 0.6 confidence
at::Tensor score_mask = scores > 0.6;
boxes = boxes.index({score_mask, Slice()});            
scores = scores.index({score_mask});            
class_indices = class_indices.index({score_mask});
```
And this gives me about 20 FPS at max but with the overlapping bounding boxes, which I can run through the NMS function but still it is 1-3 FPS for entire operation. How can I make all of this faster and optimized so that I can use this for realtime, low latency detections on a live video stream?

Even the equivalent Python program gives about 50 FPS for all these operations since that inference does the NMS and postprocessing behind the scenes. It would mean doing yolo inferences on Python would be more optimal than C++/CUDA implementation.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @shikhindahikar, thank you for your detailed question and for using Ultralytics 🚀! Your feedback about YOLOv8 LibTorch C++ inference and real-time detection performance is very valuable.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us investigate further. This will allow our team to better understand the slowdown you're experiencing with NMS and offer more targeted suggestions.

If your inquiry is about optimizing custom inference workflows or integrating YOLO into real-time applications, it would be helpful to share:
- Specific code snippets you are using for NMS and postprocessing
- Hardware specs and environment details
- Any timing or profiling logs

Our [Docs](https://docs.ultralytics.com/) and especially the [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples may also provide insights into how optimizations are managed in Python.

For additional help or real-time chat, join us on [Discord](https://discord.com/invite/ultralytics) 🎧. You can also exchange ideas on [Discourse](https://community.ultralytics.com/) or connect with others on our [Subreddit](https://reddit.com/r/Ultralytics).

## Upgrade

Please ensure you are running the latest `ultralytics` package and all up-to-date [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in these verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM ([GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/))
- **Amazon** Deep Learning AMI ([AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/))
- **Docker Image** ([Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/)) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) across macOS, Windows, and Ubuntu every 24 hours as well as on every commit.

This is an automated response to help you get started quickly. An Ultralytics engineer will review your question and provide further assistance soon!

**Reply by glenn-jocher** [reply]
The performance difference you're experiencing is likely due to inefficient NMS implementation in your C++ code compared to the optimized operations in the Python version. Here are a few approaches to improve your real-time performance:

Consider implementing a confidence threshold filter before NMS to reduce the number of detections processed - start with a higher threshold like 0.7-0.8 instead of 0.6. You can also limit `max_det` to a smaller number (like 100-300 instead of the default 300) and reduce `max_nms` from 30000 to something like 5000-10000 based on your use case.

For the NMS implementation itself, ensure you're using optimized tensor operations and consider batching multiple frames if your application allows for slight latency. The Python version benefits from highly optimized backend operations that might not be replicated in your current C++ implementation.

You might also want to experiment with the `agnostic_nms` parameter and adjust `iou_thres` - sometimes a slightly higher IoU threshold (0.5-0.6) can reduce computation while maintaining acceptable results for real-time applications.

Have you tried profiling your C++ code to identify exactly where the bottleneck occurs within the NMS operation? This could help pinpoint whether it's the sorting, IoU computation, or tensor indexing that's causing the slowdown.

**Reply by shikhindahikar** [reply]
> The performance difference you're experiencing is likely due to inefficient NMS implementation in your C++ code compared to the optimized operations in the Python version

I am using the torchvision CUDA implementation of the NMS as linked above, which is supposed to be quite fast.

> Consider implementing a confidence threshold filter before NMS to reduce the number of detections processed - start with a higher threshold like 0.7-0.8 instead of 0.6.

I have tried this above using the torch tensor functions(example in the post) and for some reason it gives me a massive decrease in performance. I have even tried doing the same confidence filtering by creating CUDA function to do that manually, still it is comparable to the tensor function result. 

You have suggested, `max_nms` and `max_det` parameter tuning. I can see these in the [ultralytics documentation](https://docs.ultralytics.com/reference/utils/ops/#ultralytics.utils.ops.non_max_suppression) for NMS in `ultralytics.utils.ops.non_max_suppression` but how can I do the same for C++? Could you please point me to a documentation?

**Reply by glenn-jocher** [reply]
For C++ LibTorch implementation, you'll need to manually implement the `max_nms` and `max_det` logic since these are Python-specific parameters. Here's how you can replicate them:

Before applying NMS, limit your candidates by sorting detections by confidence and keeping only the top `max_nms` detections (e.g., 5000-10000):

```cpp
// Sort by confidence and keep top max_nms detections
auto sorted_indices = std::get<1>(scores.sort(-1, true));
auto top_indices = sorted_indices.slice(0, 0, std::min(max_nms, (int)sorted_indices.size(0)));
boxes = boxes.index({top_indices});
scores = scores.index({top_indices});
class_indices = class_indices.index({top_indices});
```

After NMS, limit final detections to `max_det`:

```cpp
// After NMS, limit to max_det detections
if (keep.size(0) > max_det) {
    keep = keep.slice(0, 0, max_det);
}
```

The massive performance decrease you're seeing with tensor filtering might be due to memory allocation overhead. Try pre-allocating tensors or using in-place operations where possible. Also, ensure your CUDA tensors stay on GPU throughout the pipeline to avoid CPU-GPU transfer costs.

Since you're getting good performance with raw inference, the bottleneck is definitely in the post-processing. Consider implementing a simplified NMS that trades some accuracy for speed, or look into using TensorRT for end-to-end optimized inference with built-in NMS.

**Reply by shikhindahikar** [reply]
Thanks. I am working on a TensorRT implementation for this since couple of days.

**Reply by glenn-jocher** [reply]
Great choice! TensorRT should resolve your performance bottlenecks since it includes highly optimized built-in NMS operations that eliminate the post-processing overhead you were experiencing. The end-to-end optimization will likely get you much closer to your Python performance targets for real-time inference.

**Reply by shikhindahikar** [reply]
I am hoping to get better performance than Python. Will keep this thread active for next week as well, in case I want to ask more related questions.

**Reply by glenn-jocher** [reply]
You should definitely achieve better performance than Python with TensorRT - the optimized GPU kernels and fused operations typically provide significant speedup over the Python implementation. Feel free to drop any questions here as you work through the TensorRT integration, happy to help with any implementation details or performance tuning!

**Comment by Y-T-G** [comment]
You can export to TensorRT with `nms=True` and `half=True` which would include NMS within the model.

https://www.reddit.com/r/Ultralytics/comments/1jwl6se/ultralytics_postprocessing_guide/

**Reply by shikhindahikar** [reply]
Thanks a lot! I had implemented the TensorRT CUDA inference program for yolo but I am stuck at the part where I need to create NMS layer in TensorRT for using its inbuilt INMS feature. This might be helpful for me.

**Reply by glenn-jocher** [reply, marked as answer]
Perfect! That's exactly what you need - exporting with `nms=True` will embed the NMS operations directly into the TensorRT engine, eliminating the need to manually implement NMS layers. This approach gives you the end-to-end optimized pipeline you were looking for:

```python
model.export(format="engine", nms=True, half=True)
```

This will handle all the post-processing internally and should deliver the real-time performance you're targeting without the complexity of custom TensorRT NMS layer implementation. Much cleaner solution than building the NMS components from scratch!

**Reply by shikhindahikar** [reply]
Thanks! That worked. I actually first converted my `pt` model to `onnx` format with `nms=True` using the yolo CLI. After that using `trtexec` available in the TensorRT I exported it in `engine` format.

**Reply by glenn-jocher** [reply]
Excellent! That's a great alternative workflow - using the CLI to export to ONNX with `nms=True` then `trtexec` for the final engine conversion gives you fine-grained control over the TensorRT optimization parameters. This two-step approach often provides more flexibility for custom deployment scenarios. Glad you got it working and hope you're seeing the real-time performance improvements you were targeting!

**Comment by shikhindahikar** [comment]
Hi @glenn-jocher a quick update and question for you. Me and my teammates tried to compare the accuracy of Python detection and TensorRT detection against the ground truth of a video and it seems that TensorRT is giving poorer precision and recall in comparison to Python counterpart. Do you know how I can make it at least comparable accuracy? 
Both are running on fp16 precision. I have generated my engine by the steps I mentioned above earlier.

**Reply by Y-T-G** [reply]
What was your export command?

**Reply by shikhindahikar** [reply]
`yolo export model=detector.pt format=onnx nms=True half=True` for getting onnx followed by 
`trtexec --onnx=detector.onnx --saveEngine=detector.engine --fp16 --device=0 --verbose` for generating the engine. Could there be some IoU issue here internally?

**Reply by Y-T-G** [reply]
How did you check the metrics for the TensorRT model? Did you use Ultralytics?

**Reply by shikhindahikar** [reply]
No, we have a video for which we have all the frames manually annotated. Using python we get more precise results by comparing to those annotations. 
How can I use Ultralytics tools or utilities for generating metrics?

**Reply by Y-T-G** [reply]
So do you have a custom inference code for TensorRT that doesn't use Ultralytics? If so, then the issue is probably with your pre-processing code.

**Reply by shikhindahikar** [reply]
custom code as in? I just get a frame which I convert to NCHW format for TensorRT input and I have followed everything according to the TensorRT documentation for C++ inference. My model is a yolov11 model so technically it should work and it is giving me detections quite fast but I need some accuracy which I am probably losing in internal NMS of TensorRT or somewhere.

**Reply by Y-T-G** [reply]
Custom code as in you're not using Ultralytics for inference. You can run inference with TensorRT with Ultralytics directly.

Since it's custom code, then I assume it's probably an issue with your code, particularly preprocessing. Most probably not using the correct resizing method or padding.

**Reply by shikhindahikar** [reply]
I did not find any information online about running TensorRT with Ultralytics online for C++/CUDA. I only found for Python. Could you please point me to the documentation for this?

**Reply by Y-T-G** [reply]
I am referring to running prediction with TensorRT with Ultralytics in Python. If you're using a custom inference code, then you need to make sure the preprocessing is correct. This example has the correct preprocessing code in C++:

https://github.com/ultralytics/ultralytics/blob/main/examples/YOLOv8-LibTorch-CPP-Inference/main.cc

**Reply by shikhindahikar** [reply]
I am only not doing letterboxing exactly same as given here. Rest everything is the same. Will try doing letterboxing like this to see if it generates better results.

**Reply by glenn-jocher** [reply]
Letterboxing is indeed critical for maintaining accuracy - it ensures your input preprocessing matches exactly what the model expects during training. The [letterboxing implementation in the LibTorch example](https://github.com/ultralytics/ultralytics/blob/main/examples/YOLOv8-LibTorch-CPP-Inference/main.cc) preserves aspect ratio and applies proper padding, which is essential for detection accuracy.

Make sure your letterboxing maintains the same stride (32 for YOLO11) and uses the same padding values and normalization as the Python implementation. Even small differences in how you handle image resizing, padding, or pixel value scaling can cause noticeable accuracy drops with TensorRT engines.

You can also verify your preprocessing is correct by comparing intermediate tensor values between your C++ implementation and a Python inference on the same image before it hits the TensorRT engine.

**Reply by shikhindahikar** [reply]
Got it. Thanks guys! Will have to see how I can do fast letterboxing. 

I am working on a streaming software, the colour space is UYVY. So I have to first probably convert UYVY frame to RGB, then do the letterboxing and then convert it to NCHW for TensorRT input. Will probably reimplement the libtorch example letterbox function in CUDA. 

Please do let me know if there is any resource online for the letterbox algorithm you guys are using at Ultralytics.

**Reply by glenn-jocher** [reply]
The letterboxing algorithm is implemented in the [Ultralytics utils ops module](https://docs.ultralytics.com/reference/utils/ops/#ultralytics.utils.ops.letterbox) - you can find the exact implementation details there. The key is matching the `stride`, `auto` padding behavior, and pixel normalization that Ultralytics uses.

For your streaming pipeline with UYVY, implementing the entire preprocessing chain in CUDA makes sense for performance. You'll want to ensure your CUDA letterboxing matches the `center=True`, `scaleFill=False` behavior from the Python implementation to maintain accuracy parity.

The [Python letterbox function](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/utils/ops.py) source code will give you the exact algorithm parameters and logic flow to replicate in your CUDA implementation.
